### PR TITLE
Recurrent mechanic

### PR DESCRIPTION
**Pull request by timcheck**
Issue Number: #102

Objective of pull request:
Integrate Sumit's custom recurrent computation to accelerate PyTorch forward and backward computation for recurrent SNNs.

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava-dl/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [x] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-Recurrent SNN training speed scales with network size poorly

## What is the new behavior?

-Increased recurrent SNN training speed; training speed scales more favorably 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/lava/lib/dl/slayer/block/base.py [hunk 1]
```diff
@@ -1428,18 +1430,21 @@ def forward(self, x):
...
 
         if z.shape[0] == self.spike_state.shape[0]:
             spike = spike + self.spike_state
 
-        for time in range(z.shape[-1]):
-            dendrite = z[..., time:time + 1]
-            feedback = self.recurrent_synapse(spike.reshape(dendrite.shape))
-            spike = self.neuron(dendrite + feedback)
-            x[..., time:time + 1] = spike
+        ###
```
Comment (bamsumit) [inline src/lava/lib/dl/slayer/block/base.py:1439]: You can remove the commented code.

File: src/lava/lib/dl/slayer/utils/recurrent.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        grad_spike = 0
+        # grad_recurrent_mat = 
+        # torch.zeros_like(ctx.recurrent_mat).to(grad_x.device)
+        for time in range(grad_x.shape[-1])[::-1]:
+            grad_spike = grad_spike + grad_x[..., time: time + 1]
+            torch.autograd.backward(ctx.spikes[time], grad_spike)
+            grad_dend_sum = ctx.dend_sums[time].grad
+            grad_feedback = grad_dend_sum
+            grad_dendrite = grad_dend_sum
+            # if time >= 1:
```
Comment (bamsumit) [inline src/lava/lib/dl/slayer/utils/recurrent.py:97]: Remove this portion of commented code.

File: tests/lava/lib/dl/slayer/utils/test_recurrent.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        if verbose is True:
+            print("testing recurrent")
+
+        torch.manual_seed(298101)
+
+        batch_size = 2
+        n_time_steps = 1024
+        n_neur = 100
+
+        device = torch.device("cuda")
```
Comment (bamsumit) [inline tests/lava/lib/dl/slayer/utils/test_recurrent.py:31]: Remove this. Our CI does not have GPU, that's why it's failing.

**Review by bamsumit — COMMENTED** [review]
Looks good @timcheck . Please fix the comments and CI issues.

**Review by bamsumit — APPROVED** [review]

File: src/lava/lib/dl/slayer/block/base.py (hunk 1 above)
Comment (mgkwill) [inline src/lava/lib/dl/slayer/block/base.py:1439]: Generally we wouldn't merge commented code unless it is explanation code that is part of a larger comment.

**Review by mgkwill — APPROVED** [review]

**Pull request merged**